src/numerics/funcs.cpp

```diff
@@ -27,4 +29,66 @@ doublereal linearInterp(doublereal x, const vector_fp& xpts,
     return ff;
 }
 
+double trapezoidal(const Eigen::VectorXd& f, const Eigen::VectorXd& x)
+{
+    // check length
+    if (f.size() != x.size()) {
+        throw CanteraError("trapezoidal",
+                           "Vector lengths need to be the same.");
+    }
+    // Vector of f(i+1) + f(i)
+    Eigen::VectorXd f_av = f.tail(f.size() - 1) + f.head(f.size() - 1);
+    // Vector of x(i+1) - x(i)
+    Eigen::VectorXd x_diff = x.tail(x.size() - 1) - x.head(x.size() - 1);
+    // check if the coordinate is a monotonically increase vector.
+    if ((x_diff.array() <= 0.0).any()) {
+        throw CanteraError("trapezoidal",
+            "x (coordinate) needs to be the monotonically increasing.");
+    }
+    return f_av.dot(x_diff) / 2.0;
+}
+
+double simpsonQuadrature(const Eigen::VectorXd& f, const Eigen::VectorXd& x)
+{
+    if (f.size() != x.size()) {
+        throw CanteraError("simpsonQuadrature",
+                           "Vector lengths need to be the same.");
+    }
+    size_t N = f.size();
+    size_t ns = (N - 1) / 2;
+    size_t ms = (N - 1) % 2;
+    double sum = 0.0;
+    for (size_t i = 0; i < ns; i++) {
+        double c[3];
+        double xs[3] = {x[2*i],x[2*i+1],x[2*i+2]};
+        double fs[3] = {f[2*i],f[2*i+1],f[2*i+2]};
+        polyfit(3, 2, xs, fs, nullptr, c);
```

Comment:
Is it necessary to use polyfit here? The [Wikipedia article on Simpson's rule](https://en.wikipedia.org/wiki/Simpson%27s_rule) gives a direct formula for calculating it when the intervals are uneven.